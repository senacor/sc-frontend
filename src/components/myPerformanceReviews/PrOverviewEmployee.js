import React, { useEffect, useState, useContext } from 'react';
import PerformanceReviewTable from '../humanResources/PerformanceReviewTable';
import RequestPerformanceReview from './RequestPerformanceReview';
import FILTER_GROUPS from '../humanResources/filterGroups';
import PerformanceReviewTableService from '../humanResources/PerformanceReviewTableService';
import Paper from '@material-ui/core/Paper/Paper';
import Grid from '@material-ui/core/Grid';
import TableColumnSelectorMenu from '../humanResources/TableColumnSelectorMenu';
import { injectIntl } from 'react-intl';
import { getFilterPossibilities } from '../../actions/calls/filter';
import { fetchFilteredPrs } from '../../actions/calls/pr';
import CircularProgress from '@material-ui/core/CircularProgress';
import { withStyles } from '@material-ui/core';
import { ErrorContext } from '../App';

const styles = theme => ({
  ...theme,
  paper: {
    margin: theme.spacing.unit * 3
  }
});

export const PrOverviewEmployee = props => {
  const [filter, setFilter] = useState({});
  const [state, setState] = useState({
    columnsToView: null
  });

  const [filterPossibilities, setFilterPossibilities] = useState({});
  const [isLoading, setIsLoading] = useState(false);
  const [data, setData] = useState([]);

  let errorContext = useContext(ErrorContext.context);

  const { classes } = props;

  useEffect(() => {
    getFilterPossibilities(setIsLoading, setFilterPossibilities, errorContext);
  }, []);

  useEffect(
    () => {
      fetchFilteredPrs(
        filter,
        FILTER_GROUPS.EMPLOYEE,
        setData,
        setIsLoading,
        errorContext
      );
    },
    [filter]
  );

  const getColumnDefinitions = () => {
    const prTableService = new PerformanceReviewTableService(
      filterPossibilities,
      filter,
      setFilter
    );

    return [
      prTableService.employee(false),
      prTableService.deadline(),
      prTableService.occasion(props.intl),
      prTableService.projectCst(),
      prTableService.competence(props.intl),
      prTableService.level(),
      prTableService.supervisor(),
      prTableService.reviewer(),
      prTableService.result(props.intl),
      prTableService.employeePreparation(props.intl),
      prTableService.reviewerPreparation(props.intl),
      prTableService.meeting(),
      prTableService.finalState()
    ];
  };

  const getColumnsFromLocalStorage = () => {
    const columnsChecked = JSON.parse(localStorage.getItem('columnsChecked'));
    const allColumns = getColumnDefinitions();
    let result = [];

    columnsChecked.forEach((column, index) => {
      if (column) {
        result.push(allColumns[index]);
      }
    });
    return result;
  };

  const getSelectorContent = () => {
    let columns = getColumnDefinitions();
    let result = [];
    columns.forEach(column => {
      result.push({ label: column.label, value: column });
    });
    return result;
  };

  const handleChange = content => {
    setState({ columnsToView: content });
  };

  if (isLoading) {
    return <CircularProgress />;
  }

  if (!filterPossibilities.levels) {
    return null;
  }

  const { columnsToView } = state;
  let columns = columnsToView ? columnsToView : getColumnsFromLocalStorage();
  if (columns.length === 0) {
    columns = getColumnDefinitions();
  }

  return (
    <Paper className={classes.paper}>
      <Grid
        container
        direction={'row'}
        justify={'flex-end'}
        alignItems={'center'}
      >
        <Grid item>
          <RequestPerformanceReview />
        </Grid>
        <Grid item>
          <TableColumnSelectorMenu
            onChange={handleChange}
            content={getSelectorContent()}
          />
        </Grid>
      </Grid>

      <PerformanceReviewTable
        columnDefinition={columns}
        orderBy={0}
        data={data}
      />
    </Paper>
  );
};

export default injectIntl(withStyles(styles)(PrOverviewEmployee));
