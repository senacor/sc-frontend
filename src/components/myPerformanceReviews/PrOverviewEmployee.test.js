import React from 'react';
import { PrOverviewEmployee } from './PrOverviewEmployee';
import { createShallow } from '@material-ui/core/test-utils';

describe('PrOverviewReviewer Component', () => {
  let shallow = createShallow({ dive: false });

  it('should match snapshot', () => {
    let component = shallow(<PrOverviewEmployee />);

    expect(component).toMatchSnapshot();
  });
});
