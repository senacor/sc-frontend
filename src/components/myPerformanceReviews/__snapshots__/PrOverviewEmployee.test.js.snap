// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrOverviewReviewer Component should match snapshot 1`] = `
<WithStyles(Paper)>
  <WithStyles(Grid)
    alignItems="center"
    container={true}
    direction="row"
    justify="flex-end"
  >
    <WithStyles(Grid)
      item={true}
    >
      <InjectIntl(Connect(Component)) />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
    >
      <InjectIntl(TableColumnSelectorMenu)
        content={
          Array [
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.employee"
                values={Object {}}
          />,
              "value": Object {
                "label": <FormattedMessage
                  id="performancereviewtableservice.employee"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.duedate"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="deadline"
                  filterGroup="EMPLOYEE"
          >
                  <InjectIntl(Connect(WithStyles(Component))) />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.duedate"
                  values={Object {}}
          />,
                "numeric": true,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.reason"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="occasion"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "Auf Nachfrage": "ON_DEMAND",
                                          "Ende der Probezeit": "END_PROBATION",
                                          "Jährlich": "YEARLY",
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.reason"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.projectcst"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="projectCst"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "ACD_BOOT": "ACD_BOOT",
                                          "DEV_CAMP": "DEV_CAMP",
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.projectcst"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.devcon"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="competence"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "Con.": "CONSULTING",
                                          "Dev.": "DEVELOPMENT",
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.devcon"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.level"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="level"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "1": "1",
                                          "2": "2",
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.level"
                  values={Object {}}
          />,
                "numeric": true,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.supervisor"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="supervisor"
                  filterGroup="EMPLOYEE"
          >
                  <InjectIntl(Connect(EmployeeFilter)) />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.supervisor"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.reviewer"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="reviewer"
                  filterGroup="EMPLOYEE"
          >
                  <InjectIntl(Connect(EmployeeFilter)) />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.reviewer"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.rating"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="overallAssessment"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "erfüllt": "3",
                                          "nicht erfüllt": "1",
                                          "zT. nicht erfüllt": "2",
                                          "zT. übererfüllt": "4",
                                          "übererfüllt": "5",
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.rating"
                  values={Object {}}
          />,
                "numeric": true,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.employeefilled"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="employeePreparationDone"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "ja": true,
                                          "nein": false,
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.employeefilled"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.reviewerfilled"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="reviewerPreparationDone"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "ja": true,
                                          "nein": false,
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.reviewerfilled"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.termin"
                values={Object {}}
          />,
              "value": Object {
                "label": <FormattedMessage
                  id="performancereviewtableservice.termin"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
            Object {
              "label": <FormattedMessage
                id="performancereviewtableservice.final"
                values={Object {}}
          />,
              "value": Object {
                "filter": <Connect(WithStyles(Component))
                  filterBy="inProgress"
                  filterGroup="EMPLOYEE"
          >
                  <Connect(Component)
                          content={
                                  Object {
                                          "beendet": false,
                                          "laufend": true,
                                        }
                          }
                  />
          </Connect(WithStyles(Component))>,
                "label": <FormattedMessage
                  id="performancereviewtableservice.final"
                  values={Object {}}
          />,
                "numeric": false,
                "render": [Function],
                "sortValue": [Function],
              },
            },
          ]
        }
        onChange={[Function]}
      />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
  <InjectIntl(WithStyles(PerformanceReviewTable))
    columnDefinition={
      Array [
        Object {
          "label": <FormattedMessage
            id="performancereviewtableservice.employee"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="deadline"
            filterGroup="EMPLOYEE"
      >
            <InjectIntl(Connect(WithStyles(Component))) />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.duedate"
            values={Object {}}
      />,
          "numeric": true,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="occasion"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "Auf Nachfrage": "ON_DEMAND",
                              "Ende der Probezeit": "END_PROBATION",
                              "Jährlich": "YEARLY",
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.reason"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="projectCst"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "ACD_BOOT": "ACD_BOOT",
                              "DEV_CAMP": "DEV_CAMP",
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.projectcst"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="competence"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "Con.": "CONSULTING",
                              "Dev.": "DEVELOPMENT",
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.devcon"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="level"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "1": "1",
                              "2": "2",
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.level"
            values={Object {}}
      />,
          "numeric": true,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="supervisor"
            filterGroup="EMPLOYEE"
      >
            <InjectIntl(Connect(EmployeeFilter)) />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.supervisor"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="reviewer"
            filterGroup="EMPLOYEE"
      >
            <InjectIntl(Connect(EmployeeFilter)) />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.reviewer"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="overallAssessment"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "erfüllt": "3",
                              "nicht erfüllt": "1",
                              "zT. nicht erfüllt": "2",
                              "zT. übererfüllt": "4",
                              "übererfüllt": "5",
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.rating"
            values={Object {}}
      />,
          "numeric": true,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="employeePreparationDone"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "ja": true,
                              "nein": false,
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.employeefilled"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="reviewerPreparationDone"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "ja": true,
                              "nein": false,
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.reviewerfilled"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "label": <FormattedMessage
            id="performancereviewtableservice.termin"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
        Object {
          "filter": <Connect(WithStyles(Component))
            filterBy="inProgress"
            filterGroup="EMPLOYEE"
      >
            <Connect(Component)
                  content={
                        Object {
                              "beendet": false,
                              "laufend": true,
                            }
                  }
            />
      </Connect(WithStyles(Component))>,
          "label": <FormattedMessage
            id="performancereviewtableservice.final"
            values={Object {}}
      />,
          "numeric": false,
          "render": [Function],
          "sortValue": [Function],
        },
      ]
    }
    orderBy={1}
  />
</WithStyles(Paper)>
`;
